### PR TITLE
[Narwhal] update batches resend logic; add more metrics to proposer

### PR DESCRIPTION
**Pull request by mwtian**
## Description 

This PR contains a fix and a few cleanups:
1. Resending batches in headers whenever commit round is more than 2 above uncommitted header round seems not very accurate, and can introduce unnecessary resends in some cases. Instead, only batches from uncommitted headers below the highest committed header round should get resent.
2. Add metrics for number of headers and digests that are resent.
3. Use OurDigestMessage to keep track of digest metadata, instead of a tuple.
4. Clarify FIFO ordering of sending digests from Proposer.

## Test Plan 

Not adding a new unit test for resending header and digest, because we will update this logic later.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: narwhal/primary/src/proposer.rs [hunk 1]
```diff
@@ -512,33 +515,38 @@ impl Proposer {
...
                 Some((commit_round, commit_headers)) = self.rx_committed_own_headers.recv() => {
                     // Remove committed headers from the list of pending
-                    for round in &commit_headers {
-                        self.proposed_headers.remove(round);
+                    let mut max_committed_round = 0;
+                    for round in commit_headers {
+                        max_committed_round = max_committed_round.max(round);
+                        let Some(_) = self.proposed_headers.remove(&round) else {
+                            info!("Own committed header not found at round {round}, probably because of restarts.");
+                            continue;
```
Comment (akichidis) [inline narwhal/primary/src/proposer.rs:523]: What's the intention of the `continue` here? IIUC the intention is that we want to skip processing any other committed headers as we have probably restarted - in this case you want to tag the  loop outside of `tokio::select` and do `continue` on that one.

File: narwhal/primary/src/proposer.rs [hunk 2]
```diff
@@ -512,33 +515,38 @@ impl Proposer {
...
                     }
 
                     if !retransmit_rounds.is_empty() {
                         // Add the digests to retransmit to the digests for the next header
-                        digests_to_resend.extend(&self.digests);
-                        self.digests = digests_to_resend;
+                        let num_to_resend = digests_to_resend.len();
+                        for digest in digests_to_resend.into_iter().rev() {
+                            self.digests.push_front(digest);
+                        }
```
Comment (akichidis) [inline narwhal/primary/src/proposer.rs:549]: Is the intention to prioritise on the final `self.digests` list the newer batches first? If yes, then doing a `rev` on the `digests_to_resend` and a `push_front` will result in adding the elements in an order of `older --> newer`  . So either remove the `rev` or `push_back` ?

File: narwhal/primary/src/proposer.rs [hunk 3]
```diff
@@ -547,11 +555,14 @@ impl Proposer {
...
                             "Retransmit {} batches in undelivered headers {:?} at commit round {:?}, remaining headers {}",
-                            self.digests.len(),
+                            num_to_resend,
                             retransmit_rounds,
                             commit_round,
                             self.proposed_headers.len()
                         );
+
+                        self.metrics.proposer_resend_headers.inc_by(retransmit_rounds.len() as u64);
+                        self.metrics.proposer_resend_batches.inc_by(num_to_resend as u64);
```
Comment (akichidis) [inline narwhal/primary/src/proposer.rs:565]: Great new metrics 👍

File: narwhal/primary/src/proposer.rs [hunk 4]
```diff
@@ -30,7 +30,7 @@ pub struct OurDigestMessage {
     pub worker_id: WorkerId,
     pub timestamp: TimestampMs,
     /// A channel to send an () as an ack after this digest is processed by the primary.
-    pub ack_channel: oneshot::Sender<()>,
+    pub ack_channel: Option<oneshot::Sender<()>>,
```
Comment (akichidis) [inline narwhal/primary/src/proposer.rs:33]: Why an `Option` ?

File: narwhal/primary/src/proposer.rs [hunk 5]
```diff
@@ -83,11 +83,12 @@ pub struct Proposer {
     /// Holds the certificate of the last leader (if any).
     last_leader: Option<Certificate>,
     /// Holds the batches' digests waiting to be included in the next header.
-    digests: Vec<(BatchDigest, WorkerId, TimestampMs)>,
+    /// Digests are popped in LIFO order from the front.
+    digests: VecDeque<OurDigestMessage>,
```
Comment (akichidis) [inline narwhal/primary/src/proposer.rs:87]: So whenever we retry we always want to retry the "latest/newest" batches first - meaning the ones from the higher rounds, correct?

**Review by akichidis — APPROVED** [review]
Thanks @mwtian ! Do we have metrics of the current impact of the previous implementation? How did we discover the need for improvement? 


> Not adding a new unit test for resending header and digest, because we will update this logic later.

Are you looking to a specific immediate/follow up refactoring? If yes then could you please share the plan? If that's the case then I am ok adding tests then - otherwise let's add something as part of this PR.

File: narwhal/primary/src/proposer.rs (hunk 5 above)
Comment (mwtian) [reply narwhal/primary/src/proposer.rs:87]: Sorry this is a leftover comment from a closed PR. Using LIFO was the previous idea, but I think it needs to be paired with updates to error handling and back pressure logic to work. So keeping the FIFO policy.

**Comment by mwtian** [comment]
> Thanks @mwtian ! Do we have metrics of the current impact of the previous implementation? How did we discover the need for improvement?

This was discovered while working on weak links. I can ran an experiment to verify the impact.

> 
> > Not adding a new unit test for resending header and digest, because we will update this logic later.
> 
> Are you looking to a specific immediate/follow up refactoring? If yes then could you please share the plan? If that's the case then I am ok adding tests then - otherwise let's add something as part of this PR.

Switching to weak links would affect this logic. But I can try to add a test for now.

**Pull request merged**